packages/lazy-import/lib/index.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+const npa = require( 'npm-package-arg' );
+const semver = require( 'semver' );
+const execa = require( 'execa' );
+const { join } = require( 'path' );
+const { createHash } = require( 'crypto' );
+
+/**
+ * @typedef WPLazyImportOptions
+ *
+ * @property {()=>void} onInstall Callback to invoke when install starts.
+ */
+
+/**
+ * Returns an md5 hash of the given string.
+ *
+ * @param {string} text Text for which to generate hash.
+ *
+ * @return {string} md5 hash of string.
+ */
+function md5( text ) {
+	return createHash( 'md5' )
+		.update( text )
+		.digest( 'hex' );
+}
+
+/**
+ * Given an arg string as would be passed to `npm install`, returns a string to
+ * use as the alias name of the package to install locally if needed.
+ *
+ * @param {string} arg A string that you might pass to `npm install`.
+ *
+ * @return {string} Module to use as local package alias.
+ */
+function getLocalModuleName( arg ) {
+	return '@wordpress/lazy-import.' + md5( arg );
+}
+
+/**
+ * Installs npm package by arg name at the given prefix path. Creates the prefix
+ * path if it doesn't yet exist.
+ *
+ * @param {string} arg   Package install arg (name and version specification).
+ * @param {string} alias Alias to use for local module name.
+ *
+ * @return {Promise<void>} Promise resolving once package is installed.
+ */
+async function install( arg, alias ) {
+	await execa( 'npm', [ 'install', '--no-save', alias + '@npm:' + arg ] );
+}
+
+/**
+ * Given an arg string as would be passed to `npm install`, requires the package
+ * corresponding to the name and specifier parsed from the arg. If the package
+ * is not installed (or not installed at that version), the package is installed
+ * to a temporary directory at `node_modules/.wp-lazy` relative to where the
+ * current working directory.
+ *
+ * @param {string}                       arg       A string that you might pass
+ *                                                 to `npm install`.
+ * @param {Partial<WPLazyImportOptions>} [options] Optional options object.
+ *
+ * @return {Promise<NodeRequire>} Promise resolving to required module.
+ */
+async function lazyImport( arg, options = {} ) {
+	const { rawSpec, name } = npa( arg );
+
+	if ( ! name ) {
+		throw new TypeError(
+			`Unable to parse package name from \`${ arg }\`.`
+		);
+	}
+
+	const localModule = getLocalModuleName( arg );
```

Comment:
If I follow it correctly, there is a tiny probability that you could install the same version of the package several times using different aliases when you use the following syntaxes:
- `my-package`
- `my-package@1.0.0`
- `my-package@^1.0.0`
- `my-package@~1.0.0.`
and they all resolve to the same version.

I don't see it as a blocker for v1, but I was curious if you share my concern.

Reply:
> If I follow it correctly, there is a tiny probability that you could install the same version of the package several times using different aliases when you use the following syntaxes:

Yeah, it's a good point. Earlier in my iterations, I had wondered if there might be a way to avoid the hashing altogether by naming the folder based on whichever version the specifier resolves to. It would be nice, but the problem is that it's not clear up-front (at the time of calling `npm install`) what that version would be, except by triggering an extra network request to fetch the version ahead of the install.

There's certainly some prior art from [`bin/check-latest-npm.js`](https://github.com/WordPress/gutenberg/blob/30cd85aebe14eee995e3162f09d31f4d4786f101/bin/check-latest-npm.js#L10-L74) we could use to go this route, and it may actually be an advisable path.

One of the other things I wasn't quite as fond of is the fact that you can't tell based on the folder even what the name of the package is, let alone the version. It's only necessary because I wasn't certain if all possible characters in this argument (version specifier) are valid for folder names. But if we can know the package name and specific SemVer version, we could construct folders like:

```
node_modules/@wordpress/lazy-import.my-package.1.0.0/package.json
```

Reply:
I don't love the current implementation, but I'm having a hard time thinking if there's a good alternative.

I considered:

1. Fetch from the NPM registry beforehand? Downsides are: Extra (redundant) network request, not clear how to fetch from NPM based on a version specifier (vs. a specific version like in #21306).
2. Install and then rename the directory once complete? Downsides are: There's no easy way to check for the local temporary install to avoid `npm install` on subsequent runs, since the specific desired version can't be known. It might technically be possible to find all locally-installed versions and run them against the version specifier using `semver` package.

Reply:
I agree that using an additional network request is concerning here. Well, if you use consistently one way to lazy import it won’t be an issue. I have just realized that the same happens when using Lerna and monorepo setup 😅